rekall/plugins/addrspaces/pmem.py

```diff
@@ -0,0 +1,341 @@
+    "EfiMemoryMappedIO": "",  # ACPI tables.
+    "EfiMemoryMappedIOPortSpace": "",  # ACPI tables.
+    "EfiPalCode": "r",  # OS-dependent. Largely read-only.
+    "EfiMaxMemoryType": "rw",  # No idea (adamsh). Looks like general use?
+}
+
+
+def efi_type_writable(efi_type):
+    return "w" in EFI_SEGMENTS_SAFETY[str(efi_type)]
+
+
+def efi_type_readable(efi_type):
+    return "r" in EFI_SEGMENTS_SAFETY[str(efi_type)]
+
+
+# Adapted from http://wiki.phoenix.com/wiki/index.php/EFI_MEMORY_DESCRIPTOR
+#
+# Every kernel that supports EFI will have a similar struct, but called
+# something else. For example, on XNU this is called EfiMemoryRange, but
+# XNU's definition doesn't account for the trailing 8-byte padding.
+EFI_VTYPES = {
+    "EFI_MEMORY_DESCRIPTOR": [0x30, {
+        "Type": [0x0, ["Enumeration", dict(
+            choices={
+                # See typedef here:
+                # http://wiki.phoenix.com/wiki/index.php/EFI_MEMORY_TYPE
+                0: "EfiReservedMemoryType",
+                1: "EfiLoaderCode",
+                2: "EfiLoaderData",
+                3: "EfiBootServicesCode",
+                4: "EfiBootServicesData",
+                5: "EfiRuntimeServicesCode",
+                6: "EfiRuntimeServicesData",
+                7: "EfiConventionalMemory",
+                8: "EfiUnusableMemory",
+                9: "EfiACPIReclaimMemory",
+                10: "EfiACPIMemoryNVS",
+                11: "EfiMemoryMappedIO",
+                12: "EfiMemoryMappedIOPortSpace",
+                13: "EfiPalCode",
+                14: "EfiMaxMemoryType"},
+            target="unsigned int")]],
+        "Pad": [0x4, ["unsigned int"]],
+        "PhysicalStart": [0x8, ["unsigned long"]],
+        "VirtualStart": [0x10, ["unsigned long"]],
+        "NumberOfPages": [0x18, ["unsigned long"]],
+        "Attribute": [0x20, ["unsigned long"]],
+        "Pad2": [0x28, ["unsigned long"]]}]}
+
+
+class EFIProfile(basic.ProfileLP64, basic.BasicClasses):
+    """Profile for EFI types. Used for staging."""
+
+    @classmethod
+    def Initialize(cls, profile):
+        super(EFIProfile, cls).Initialize(profile)
+        profile.add_types(EFI_VTYPES)
+
+
+class PmemAddressSpace(addrspace.CachingAddressSpaceMixIn,
+                       addrspace.RunBasedAddressSpace):
+    """Address space specific to the pmem device."""
+
+    __name = "pmem"
+    order = 1
+    __image = True
+
+    def __init__(self, base=None, filename=None, **kwargs):
+        self.as_assert(base == None,
+                       "Must be mapped directly over a raw device.")
+        super(PmemAddressSpace, self).__init__(**kwargs)
+        self.phys_base = self
+
+        path = filename or (self.session and self.session.GetParameter(
+            "filename"))
+
+        self.as_assert(path, "Filename must be specified.")
+        self.fname = path
+
+        self.as_assert(path.startswith("/dev/"), "Pmem must be a device.")
```

Comment:
not strictly needed - we can just bail out if the ioctl fails. (A node file can exist anywhere - not necessarily on /dev/).